Minio/DataModel/ServerSideEncryption.cs

```diff
@@ -0,0 +1,162 @@
+/*
+ * Minio .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using System;
+using System.Text;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+
+using Minio.Helper;
+
+namespace Minio.DataModel
+{
+  // Type of server side encryption
+   public enum EncryptionType {
+     SSE_C,
+     SSE_S3,
+     SSE_KMS
+   }
+
+   // ServerSideEncryption interface
+   public interface ServerSideEncryption
+   {
+    // GetType() needs to return the type of Server Side encryption
+    EncryptionType GetType() ;
+    // Marshals the server side encryption headers into dictionary
+    void Marshal(Dictionary<string,string> headers) ;
+   }
+
+   // Server side encryption with customer provided keys (SSE-C)
+    public class SSEC : ServerSideEncryption {
+    
+    // secret AES-256 Key
+    protected string key;
+      public new EncryptionType GetType()
+      {
+        return EncryptionType.SSE_C;
+      }
+
+      public virtual void Marshal(Dictionary<string,string> headers) 
+      {
+        var md5SumStr = utils.getMD5SumStr(this.key);
+        headers.Add("X-Amz-Server-Side-Encryption-Customer-Algorithm", "AES256");
+        headers.Add("X-Amz-Server-Side-Encryption-Customer-Key",key)  ;
```

Comment:
This needs to be converted into a byte array using a PBKDF function just like in `minio-go`

https://github.com/minio/minio-go/blob/master/examples/s3/put-encrypted-object.go#L64

https://github.com/minio/minio-go/blob/master/pkg/encrypt/server-side.go#L62

Reply:
Agree here, assuming that user will provide 32 byte long string is not acceptable. We should provide all the versions provided by minio-go . It's a well thought out implementation.

Reply:
ok, changed implementation to accept a byte array as key, and validate length. c# has Aes class that generates keys, which is equivalent of PBKDF used in minio-go, this is documented in examples.